src/wp-admin/includes/class-wp-filesystem-ftpext.php

```diff
@@ -412,18 +412,18 @@ public function delete( $file, $recursive = false, $type = false ) {
 	 * Checks if a file or directory exists.
 	 *
 	 * @since 2.5.0
+	 * @since 6.0.0 Rewrite using ftp_rawlist, uses 'LIST' on FTP server
+	 *              takes file path or directory path as parameter.
 	 *
 	 * @param string $file Path to file or directory.
 	 * @return bool Whether $file exists or not.
 	 */
 	public function exists( $file ) {
-		$list = ftp_nlist( $this->link, $file );
-
-		if ( empty( $list ) && $this->is_dir( $file ) ) {
-			return true; // File is an empty directory.
+		if ( $this->is_dir( $file ) ) {
+			return true;
 		}
 
-		return ! empty( $list ); // Empty list = no file, so invert.
+		return ! empty( ftp_rawlist( $this->link, $file ) );
```

Comment:
This looks very close to the reverted attempt in [[33648]](https://core.trac.wordpress.org/changeset/33648) to move away from nlist. The earlier commit had a `-a`

A latter attempt was to get a directory listing of the containing folder but it ended up failing too.

Reply:
The `-a` was to grab the hidden files from the directory. In this PR the initial `$this->is_dir()` check obviates the need to bring in a directory listing. `ftp_rawlist()` accepts a file path as a parameter even though the PHP manual page doesn't report it.

In the FTP servers we tested `vsftpd` and `Filezilla Server` if the file didn't exist the return was `false` or `array()`.

Reply:
I checked out the PR and tested with PureFTP (which comes up regularly in these tickets, particularly [this one](https://core.trac.wordpress.org/ticket/34976)).

Results using `$wp_filesystem->exists()` are below.

PureFTP with `define( 'FS_METHOD', 'ftpext' );`:
```
dir: true
hidden file: true
empty hidden file: true
file: true
empty file: true
non-existent dir: false
non-existent file: false
```

PureFTP with `define( 'FS_METHOD', 'ftpsockets' );`:
```
dir: true
hidden file: true
empty hidden file: true
file: true
empty file: true
non-existent dir: false
non-existent file: false
```

Reply:
@peterwilsoncc Given the test results above, what are your thoughts?